### PR TITLE
Feature/sync terminals

### PR DESCRIPTION
**Pull request by maf-embraceit**

### CODE REVIEW

**Review by shahbaz-mehar — COMMENTED** [review]
Changes have been reviewed.

File: admin/controller/extension/module/altapay.php [hunk 1]
```diff
@@ -20,180 +21,255 @@ public function index()
...
-                            chmod($frontend_model_path, 0664);
-                        } else {
-                            $data['refresh_terminals_error'] = $this->language->get('text_altapay_terminals_error');
+        $data['refresh_terminals_error'] = '';
+        $data['sync_terminals_success'] = '';
+        $data['sync_terminals_error'] = '';
+
+        if (($this->request->server['REQUEST_METHOD'] == 'POST') && $this->validate() && isset($_POST['terminalsync'])) {
+
+            $terminals_refreshed_query = $this->db->query("SELECT `value` FROM " . DB_PREFIX . "setting WHERE `key` LIKE 'altapay_terminals_refreshed'");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:35]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 2]
```diff
@@ -20,180 +21,255 @@ public function index()
...
+
+            if ($terminals_refreshed_query->num_rows and $terminals_refreshed_query->row['value']) {
+
+                $store_country_query = $this->db->query("SELECT iso_code_2 FROM " . DB_PREFIX . "country WHERE `country_id` = " . $this->config->get('config_country_id'));
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:39]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 3]
```diff
@@ -20,180 +21,255 @@ public function index()
...
+
+                if ($store_country_query->num_rows) {
+                    $terminal_installed_query = $this->db->query("SELECT setting_id FROM " . DB_PREFIX . "setting WHERE `key` LIKE 'payment_Altapay_%' LIMIT 1");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:42]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 4]
```diff
@@ -347,4 +424,39 @@ public function uninstall()
...
+    /**
+     * @param string $code
+     * @param string $key
+     * @param integer|string $value
+     *
+     * @return void
+     */
+    public function addSettingField($code, $key, $value)
+    {
+        $query = $this->db->query("SELECT setting_id FROM " . DB_PREFIX . "setting WHERE `key` = '" . $key . "'");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:437]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 5]
```diff
@@ -347,4 +424,39 @@ public function uninstall()
...
+
+        if ($query->num_rows) {
+            $this->db->query("UPDATE " . DB_PREFIX . "setting SET `code` = '" . $code . "', `value` = '" . $value . "' WHERE `key` = '" . $key . "'");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:440]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 6]
```diff
@@ -347,4 +424,39 @@ public function uninstall()
...
+        } else {
+            $this->db->query("INSERT INTO  " . DB_PREFIX . "setting  (`store_id`,`code`,`key`, `value`,`serialized`) VALUES(0, '" . $code . "', '" . $key . "', '" . $value . "', 0)");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:442]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 7]
```diff
@@ -347,4 +424,39 @@ public function uninstall()
...
+
+    /**
+     * @param string $type
+     * @param string $code
+     *
+     * @return void
+     */
+    public function addExtensionField($type, $code)
+    {
+        $query = $this->db->query("SELECT extension_id FROM " . DB_PREFIX . "extension WHERE `code` = '" . $code . "'");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:456]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php [hunk 8]
```diff
@@ -347,4 +424,39 @@ public function uninstall()
...
+        if ($query->num_rows == 0) {
+            $this->db->query("INSERT INTO  " . DB_PREFIX . "extension  (`type`,`code`) VALUES('" . $type . "', '" . $code . "')");
```
Comment (emicha) [inline admin/controller/extension/module/altapay.php:458]: Always use prepared statements for querying, never concatenating strings

File: admin/controller/extension/module/altapay.php (hunk 1 above)
Comment (maf-embraceit) [reply admin/controller/extension/module/altapay.php:35]: [It looks like](https://github.com/opencart/opencart/blob/master/upload/system/library/db.php) OpenCart does not have support for prepared statement in their custom DB interface.

Comment (maf-embraceit) [reply admin/controller/extension/module/altapay.php:35]: @emicha I've taken measures to the maximum extent possible in OpenCart in the below commit belonging to the next pull request
https://github.com/AltaPay/plugin-opencart3/pull/17/commits/1256ccf8696af93c2b5e3784e4523d5b832e02bf

**Review by emicha — APPROVED** [review]

**Pull request merged**